File: cpp/base/src/main/vsdk/toolkit/environment/geometry/curve/Curve.cpp
#include "vsdk/toolkit/environment/geometry/curve/Curve.h"
#include "vsdk/toolkit/environment/geometry/element/Ray.h"
#include "vsdk/toolkit/environment/geometry/element/RayHit.h"
Ray* Curve::doIntersectionFirstHit(const Ray&)
{
    return nullptr;
}

bool Curve::doIntersectionFirstHit(const Ray& inRay, RayHit* outHit)
{
    Ray* hit = doIntersectionFirstHit(inRay);
    if ( hit == nullptr ) {
        return false;
    }

    if ( outHit != nullptr ) {
        outHit->setRay(*hit);
        doExtraInformation(*hit, hit->getT(), outHit);
    }
    return true;
}

void Curve::doExtraInformation(const Ray&, double, RayHit*)
{
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/curve/Curve.ts
import { Geometry, type GeometryRay, type GeometryRayHit } from "../Geometry.js";
export abstract class Curve<
    TRay extends GeometryRay = GeometryRay,
    THit extends GeometryRayHit = GeometryRayHit,
> extends Geometry<TRay, THit> {
    public doIntersectionFirstHit(_inRay: TRay, _outHit: THit): boolean {
        return false;
    }
    public override doExtraInformation(_inRay: TRay, _inT: number, _outHit: THit): void {}
}
